cache/volume_cache.go

```
@@ -0,0 +1,145 @@
+package cache
+
+import (
+	"encoding/json"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+)
+
+type VolumeCache struct {
+	dir          string
+	backupDir    string
+	stagingDir   string
+	committedDir string
+}
+
+func NewVolumeCache(dir string) (*VolumeCache, error) {
+	if _, err := os.Stat(dir); err != nil {
+		return nil, err
+	}
+
+	c := &VolumeCache{
+		dir:          dir,
+		backupDir:    filepath.Join(dir, "committed-backup"),
+		stagingDir:   filepath.Join(dir, "staging"),
+		committedDir: filepath.Join(dir, "committed"),
+	}
+
+	if err := c.setupStagingDir(); err != nil {
+		return nil, errors.Wrapf(err, "initializing staging directory '%s'", c.stagingDir)
+	}
+
+	if err := os.RemoveAll(c.backupDir); err != nil {
+		return nil, errors.Wrapf(err, "removing backup directory '%s'", c.backupDir)
+	}
+
+	if err := os.MkdirAll(c.committedDir, 0777); err != nil {
+		return nil, errors.Wrapf(err, "creating committed directory '%s'", c.committedDir)
+	}
+
+	return c, nil
+}
+
+func (c *VolumeCache) Name() string {
+	return c.dir
+}
+
+func (c *VolumeCache) SetMetadata(metadata Metadata) error {
+	metadataPath := filepath.Join(c.stagingDir, MetadataLabel)
+	file, err := os.Create(metadataPath)
+	if err != nil {
+		return errors.Wrapf(err, "creating metadata file '%s'", metadataPath)
+	}
+	defer file.Close()
+
+	if err := json.NewEncoder(file).Encode(metadata); err != nil {
+		return errors.Wrap(err, "marshalling metadata")
+	}
+
+	return nil
+}
+
+func (c *VolumeCache) RetrieveMetadata() (Metadata, error) {
+	metadataPath := filepath.Join(c.committedDir, MetadataLabel)
+	file, err := os.Open(metadataPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return Metadata{}, nil
+		}
+		return Metadata{}, errors.Wrapf(err, "opening metadata file '%s'", metadataPath)
+	}
+	defer file.Close()
+
+	metadata := Metadata{}
+	_ = json.NewDecoder(file).Decode(&metadata)
```

Same as above, don't ignore the error.

---

Same as previous comment